### ISSUE TITLE
Open file starting at line number

### ISSUE THREAD

**Issue by DanielCardonaRojas** [sharkdp/bat#1185]
First of all great work! Loving this CLI 

I checked the man page and did not see an option to open a file starting a line number, otherwise if
this feature does exist please document.

**Comment by Kienyew** [comment]
I think there's a `--line-range` option in bat, which might does what you need 😄
```man
...
       -r, --line-range <N:M>...

              Only print the specified range of lines for each file. For example:

              --line-range 30:40
                     prints lines 30 to 40

              --line-range :40
                     prints lines 1 to 40

              --line-range 40:
                     prints lines 40 to the end of the file
...
```

**Comment by DanielCardonaRojas** [comment]
Kind of does but does not let scroll to lines previous to start of range.

**Comment by sharkdp** [comment]
You can try to use `--pager=…` to pass explicit options to less. Something like:
```
bat --pager="less -FR +10" src/assets_metadata.rs
```

You will need to add an offset of 3 lines for the bat header.

**Comment by Andriamanitra** [comment]
It would be beneficial if bat let line offset arguments pass through to less (or another pager, I think all of the most common ones use similar line offset argument syntax). This would allow users to easily get syntax highlighting from bat when an external program runs a command like `$PAGER +263 /foo/bar/baz`. 

I specifically ran into this issue with Julia's [less function](https://github.com/JuliaLang/julia/blob/master/stdlib/InteractiveUtils/src/editless.jl#L245) -- as it stands you'd need to set $PAGER to quite an impressive bash spell in order to make this work.

**Comment by sontek** [comment]
@sharkdp That doesn't seem to work:

```
rg --vimgrep _files | fzf --delimiter=: --preview='bat --color always -H {2} --pager "less -RF +{2}" {1}'
```

Page never jumps down to the line needed to start at

**Comment by QuarticCat** [comment]
It seems that `--pager="less -FR +10"` cannot correctly handle wrappings. When there are wrapping lines, `less` jumps to a wrong position. This also happens when we want to set line offset in `fzf` because `fzf` doesn't know what is the actual "line". We still need a starting line number flag.

**Comment by sharkdp** [comment]
> When there are wrapping lines, `less` jumps to a wrong position. This also happens when we want to set line offset in `fzf` because `fzf` doesn't know what is the actual "line". We still need a starting line number flag.

Fair point. If we would add a feature like `--scroll-to <LINE>`, that would be `less`-specific. So far, we have avoided pager-specific features. Or do you have any other ideas on how to implement this?

**Comment by QuarticCat** [comment]
@sharkdp Can we pass the calculated line number to the `--pager` flag like `fzf`? For example, `--pager="less -FR +{l}"`.

**Comment by sharkdp** [comment]
That sounds like a cool idea. Maybe there would be more variables that could be useful:
- `{effective-scroll-line}` for the `bat`-computed LINE that needs to be passed to the `+…` command
- `{scroll-line}` for the actual user-provided `<LINE>` argument
- `{file-name}`
- `{detected-syntax}`
- …

However, how can we provide this feature by default? We would need to modify the default value for `--pager` to be something like
```
less --other --default --options +{effective-scroll-line}
```

But what if `--scroll-to` isn't used? Should we simply set `effective-scroll-line = 0`? What other side-effects would this have?

Also, this would require every bat user with a custom `--pager` command in their config to now add the new `+{effective-scroll-line}` argument to make `--scroll-to` work.

If we don't add `+{effective-scroll-line}` by default, users of `--scroll-to` would be required to specify `--pager="… +{effective-scroll-line}"` themselves.

**Comment by QuarticCat** [comment]
@sharkdp I have another proposal: make the content of `--pager` flag a shell script statement, and provide those variables as environment variables, so that we don't need to create one more DSL to deal with complex situations.

For example, `bat --pager='less -FR ${line:++$line}'`. When `$line` is set, this is equivalent to `bat --pager='less -FR +$line'`. Otherwise, this is equivalent to `bat --pager='less -FR '`. We can also give `$line` a default value `0` or `1`, which will make the default `--pager` flag simpler. If someone's pager cannot handle these values, he can use shell script to solve this problem.

As far as I know, the configuration of [`fzf-tab`](https://github.com/Aloxaf/fzf-tab) is done this way. See https://github.com/Aloxaf/fzf-tab/wiki/Configuration

> Also, this would require every bat user with a custom `--pager` command in their config to now add the new `+{effective-scroll-line}` argument to make `--scroll-to` work.

Maybe we can add some hints to the help message of `--scroll-to` to reduce confusion. Besides, if the shell that executes the pager command is fixed, it is possible to detect which variables are set but not used and prompt some warnings. But I rather not do that.

**Comment by sharkdp** [comment]
@QuarticCat Thank you. I think I also like this better.

Unfortunately, the questions from my comment still remain, right?

**Comment by sharkdp** [comment]
Maybe this also opens up some new problems though. Especially concerning cross-platform. Which shell do you call on Windows, on macOS, etc.?

**Comment by QuarticCat** [comment]
@sharkdp

> Unfortunately, the questions from my comment still remain, right?

That's right. I have no idea how to solve them.

> Which shell do you call on Windows, on macOS, etc.?

Maybe get the current shell from `$SHELL` env variable, and provide default config for common shells. For other shells, we fall back to a less powerful config that has no variable usage.

I don't know whether Windows has `$SHELL`. If not, there must be some way to determine which shell we are using.

Apart from that, users can use `$BAT_SHELL` to override this value, just as `$BAT_PAGER` and `$PAGER`.

Honestly, I don't like this solution. `fzf-tab` is a Zsh plugin so it doesn't need to consider this problem. I can't find a similar example of the situation of `bat`.

**Comment by QuarticCat** [comment]
I have a new idea. If `bat` has a utility to print the actual line number, we don't need to modify any flags. For example:

```console
$ bat --pager="sh -c 'less -FR +$(bat-line --flags --line-at=15)'" --flags
```

Users can shorten this command by defining their own functions.

I know this is absolutely not elegant. Hope it can bring you some inspiration.

**Comment by BuonOmo** [comment]
For fzf users, there is actually no need for such an option. fzf's `--preview-window` actually solves that issue very well.

To open a preview with a fixed header (the title) and the highlighted-line in sight (at the first quarter of the preview):

```bash
rg --line-number some_regex |
	fzf \
		--delimiter=: \
		--preview='bat --color=always {1} --highlight-line={2}' \
		--preview-window='~3,+{2}+3/4'
```

And [here's](https://github.com/BuonOmo/dotfiles/blob/main/.zsh/custom/functions/rgc) a bit more advanced function to open vscode with a similar command

**Comment by QtRoS** [comment]
@juanMarinero thank you for mentioning rga-fzf, it's really great! It is [coded directly in rga](https://github.com/phiresky/ripgrep-all/blob/master/src/bin/rga-fzf.rs) now as a separate executable. But how can I use **bat** as preview now? Seems there is no way...
